### PR TITLE
Fix #10870 - ChartJS: new approach

### LINKED ISSUE #5665: ChartsJS: TimePoint value type
## 1) Environment
- PrimeFaces version: 8.0

## 2) Expected behavior
Charts js allows to pass datapoints as x/y Tuples. Right now Primefaces implemented that using the NumericPoint class. Which works, but only accepts, oh wonder, numeric points ;-).
It would be greate if there would be also a DatePoint class, or something similar, which accepts, a combination of LocalDatetime and Numeric points. 
e.g.
x = 2020-01-01-12-00-00
y = 123.45

Charts Js accepts that without any problems and shows them properly when using the type "time" for an axes.

## 3) Actual behavior
I'ts not possible to pass a Date (or LocalDatetime or ...) value to chars js. I have to figure out the conversation on my own...

### LINKED ISSUE #10849: Charts: unified and consistent data interface classes
### Description

As a developer of charts that often stack multiple elements (bar, line) to show the relationship between 2 or more data sets where my number of data points is not consistently equal; I want a consistent way to feed data to the Cartesian chart types that preserve my x / y tuples so that I do not need to pre parse all my data points to get a distinct list of X axis labels, then build each data set ensuring that I provide a value for each X value I found across all datasets, but instead let chart.js figure it out for me, instead of needing to write a potentially error prone hack to massage all the data points to fit the primefaces wrapper when the underlying chart.js can handle the tuples.

### Describe the solution you would like

For each of the ChartDataSet implementations, have the data field accept a List<Object> instead of List<Number> (or a list of an abstraction that makes sense to collect x / y tuples)  like the LineChartDataSet currently does so that if a developer wants to pass in a list of NumericPoint, or a class of their own that has a toString implementation that produces JSON that will output correctly when processed by ChartUtils.writeDataValue() the same way BubblePoint and NumericPoint do.

### Additional context

The specific classes that would need to be modified to need a minimally viable fix are:  BarChartDataSet, PolarAreaChartDataSet, and RadarChartDataSet.

### LINKED ISSUE #10870: ChartJS: new approach
### Description

Based on discussions #8819 and #10849

I'll try to sum up the big lines:
- Have one and only component: `<p:chart />`. There is no need to have multiple components, at worse `<p:chart />` should redirect to the proper renderer depending the type: lucky for us chartJS does the job for us.
- Model abstraction: As of today, PF is responsible to implement chartJS API into Java. First, chartJS API is freaking big! Second, It's hard to follow chartjs model changes from one version to another, which makes us left behind compare to the version we are already using: this is frustrating for users.
- KISS:  `<p:chart />` could be a very simple JSF wrapper of chartJS, take a String as input, and handle a few ajax events out of the box. As it takes a String, users are free to either implement their own API and serialize it into JSON, or use existing solutions like https://github.com/xdev-software/chartjs-java-model (which is already 4.4 compatible and java 11 compatible since 1.0.2) 
- Deprecate current impl for 14 

### Describe the solution you would like

**XHTML**

```xml
<div class="card">
    <h5>Plain JSON</h5>
    <p:chart value="#{chartJsView.barModel}" style="width: 100%; height: 500px;">
        <p:ajax event="itemSelect" listener="#{chartJsView.itemSelect}" update="growl"/>
    </p:chart>
</div>
<div class="card">
    <h5>XDEV ChartJS model</h5>
    <p:chart value="#{chartJsView.barModel2}" style="width: 100%; height: 500px;">
        <p:ajax event="itemSelect" listener="#{chartJsView.itemSelect}" update="growl"/>
    </p:chart>
</div>
```

**JAVA**
```java
    public void createBarModel() {
        barModel = "{" +
            "    \"type\": \"bar\"," +
            "    \"data\": {" +
            "        \"datasets\": [" +
            "            {" +
            "                \"type\": \"bar\"," +
            "                \"data\": [" +
            "                    65," +
            "                    59," +
            "                    80," +
            "                    81," +
            "                    56," +
            "                    55," +
            "                    40" +
            "                ]," +
            "                \"label\": \"My First Dataset\"," +
            "                \"hidden\": false," +
            "                \"backgroundColor\": \"rgba(255, 99, 132, 0.2)\"," +
            "                \"borderColor\": \"rgb(255, 99, 132)\"," +
            "                \"borderWidth\": 1" +
            "            }," +
            "            {" +
            "                \"type\": \"bar\"," +
            "                \"data\": [" +
            "                    85," +
            "                    69," +
            "                    20," +
            "                    51," +
            "                    76," +
            "                    75," +
            "                    10" +
            "                ]," +
            "                \"label\": \"My Second Dataset\"," +
            "                \"hidden\": false," +
            "                \"backgroundColor\": \"rgba(255, 159, 64, 0.2)\"," +
            "                \"borderColor\": \"rgb(255, 159, 64)\"," +
            "                \"borderWidth\": 1" +
            "            }" +
            "        ]," +
            "        \"labels\": [" +
            "            \"January\"," +
            "            \"February\"," +
            "            \"March\"," +
            "            \"April\"," +
            "            \"May\"," +
            "            \"June\"," +
            "            \"July\"" +
            "        ]" +
            "    }," +
            "    \"options\": {" +
            "        \"responsive\": true," +
            "        \"maintainAspectRatio\": false," +
            "        \"barPercentage\": 0.9," +
            "        \"indexAxis\": \"x\"," +
            "        \"scales\": {" +
            "            \"y\": {" +
            "                \"offset\": true," +
            "                \"stacked\": false," +
            "                \"reverse\": false," +
            "                \"beginAtZero\": true," +
            "                \"ticks\": {" +
            "                    \"autoSkip\": true," +
            "                    \"mirror\": false" +
            "                }" +
            "            }" +
            "        }," +
            "        \"plugins\": {" +
            "            \"title\": {" +
            "                \"display\": true," +
            "                \"text\": \"Bar Chart\"" +
            "            }" +
            "        }" +
            "    }" +
            "}";
    }

   // using XDEV ChartJS model
    public void createBarModel2() {
        barModel2 = new BarChart()
            .setData(new BarData()
                .addDataset(new BarDataset()
                    .setData(65, 59, 80, 81, 56, 55, 40)
                    .setLabel("My First Dataset")
                    .setBackgroundColor(new Color(255, 99, 132, 0.2))
                    .setBorderColor(new Color(255, 99, 132))
                    .setBorderWidth(1))
                .addDataset(new BarDataset()
                    .setData(85, 69, 20, 51, 76, 75, 10)
                    .setLabel("My Second Dataset")
                    .setBackgroundColor(new Color(255, 159, 64, 0.2))
                    .setBorderColor(new Color(255, 159, 64))
                    .setBorderWidth(1)
                )
                .setLabels("January", "February", "March", "April", "May", "June", "July"))
            .setOptions(new BarOptions()
                .setResponsive(true)
                .setMaintainAspectRatio(false)
                .setIndexAxis(BarOptions.IndexAxis.X)
                .setScales(new Scales().addScale(Scales.ScaleAxis.Y, new BarScale < CategoryTicks > ()
                    .setBarPercentage(BigDecimal.valueOf(0.9))
                    .setStacked(false)
                    .setTicks(new CategoryTicks()
                        .setAutoSkip(true)
                        .setMirror(true))))
                .setPlugins(new Plugins()
                    .setTitle(new software.xdev.chartjs.model.options.Title()
                        .setDisplay(true)
                        .setText("Bar Chart using XDEV java model")))
            ).toJson();
    }
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rapster**
Fix #10870

### CODE REVIEW

**Comment by melloware** [comment]
This looks great but the build is failing the logs are filled with the JSON output.  Might want to scale back the logging?

**Comment by Rapster** [comment]
Ok, this happens when PFE optimizer is executed:

> [ERROR] JSC_LANGUAGE_FEATURE. This language feature is only supported for UNSTABLE mode or better: Public class fields. at D:/Work/git/primefaces/primefaces/target/classes/META-INF/resources/primefaces/chart/chart.source.js line 1 : 56272

I used UMD file: that was the only one properly running...

**Comment by Rapster** [comment]
@asdef Did you get the chance to try this?

**Comment by asdef** [comment]
@Rapster trying to get it to build and run locally now.  apologies for the delay.

**Comment by Rapster** [comment]
No worries, I believe I shared with you the jitpack coordinates to make it easier for you: https://github.com/primefaces/primefaces/pull/10903#issuecomment-1806437086

 > @asdef feel free to try with: https://jitpack.io/#Rapster/primefaces/fix-10870-SNAPSHOT (build on top of master)

**Comment by asdef** [comment]
@Rapster I'm new to jitpack, so I cloned your repo from github and created a branch fix-10870  based on your fix-10870, and updated primefaces-showcase to include the jitpack repository and dependency, but it failed.  Am I going about this the wrong way?

**Comment by Rapster** [comment]
You can either clone my repo, and build it, or use jitpack (which will do it for you, simply follow instructions in the "how to" section)

**Comment by asdef** [comment]
@Rapster I finally got thru the yak shaving exercise of being able to actually connect to jitpack (employer network security stuff), and I don't think your last run saved successfully there due to a timeout on the build which I saw in the [logs](https://jitpack.io/com/github/Rapster/primefaces/fix-10870-6_1-g0a4f88b-8399/build.log).  When I try to resolve dependencies for a new project with the jitpack dependency and repository I get a "Could not find artifact com.github.Rapster:primefaces:jar:fix-10870-SNAPSHOT in ([jitpack.io](https://jitpack.io))" error from maven.  Am I missing something obvious?

**Comment by Rapster** [comment]
Sounds like I wanted to save you some time, but went the opposite. I guess, you'll have to clone the repo and build it on your own

**Comment by asdef** [comment]
@Rapster I've built the entire project (clean install) locally on JDK 17, and attempted to deploy the showcase app on a brand new wildfly 30 install with JDK 17, and it fails looking for a j2ee component: {"WFLYCTL0080: Failed services" => {"jboss.undertow.deployment.default-server.default-host.\"/primefaces-showcase-14.0.0-SNAPSHOT\".UndertowDeploymentInfoService" => "java.lang.ClassNotFoundException: javax.faces.webapp.FacesServlet from [Module \"deployment.primefaces-showcase-14.0.0-SNAPSHOT.war\" from Service Module Loader]
    Caused by: java.lang.ClassNotFoundException: javax.faces.webapp.FacesServlet from [Module \"deployment.primefaces-showcase-14.0.0-SNAPSHOT.war\" from Service Module Loader]"}}.  I'm guessing this is my issue because I forgot to include something.  In an effort to eliminate variables, how do you run the showcase for your branch?

**Comment by Rapster** [comment]
Same as normal showcase

**Comment by asdef** [comment]
@Rapster thanks!  didn't know about the jetty option.  Got it running and attempting modifications to use the date handler.

**Comment by asdef** [comment]
@Rapster just confirmed it works for my date time x axis use case.  I did discover one potential issue, in that if a developer puts a p:chart and one of the older approaches on the same page, it breaks both charts.  if this is because you haven't removed the old implementation yet, please ignore this, if the intention was to keep both, a warning about it in documentation or a change to make them cooperate with each other should be done..

**Comment by Rapster** [comment]
Both component will have to cohabitate for some time (at least one major release), but I don't see users using both components and even less in the same page (unless the new one can't handle a scenario)

**Comment by asdef** [comment]
given how little I know about the cadence for changes in Primefaces, how fast can I reasonably expect this change to make it to a published release?  Just trying to figure out how to fit it with my product's release schedule...

**Comment by Rapster** [comment]
I don't think we mentionned anything about PF 14 release date yet, right now I don't have time for JSF, what I have left to do is doc + showcase, wanna help 😅 ?

**Comment by asdef** [comment]
Happy to help, but unsure of the best way to do that.  Should I create a PR to your feature branch on your repo clone, or is there some better approach for collaboration on a branch?

**Comment by melloware** [comment]
@Rapster should we merge this so it's easier for @asdef to submit PRs for?  We still have some time before 14 so I think it is safe to merge ?

**Comment by Rapster** [comment]
It shoult be possible to make a PR on my branch (don't want to commit on master work half way done)

**Comment by melloware** [comment]
I am working on this. @asdef can you paste a snippet of how you did the Time Axis?  I will create a showcase example for others.

**Comment by melloware** [comment]
@Rapster I created a different PR based on yours . Sorry it was just easier for me to work with but all docs and examples updated as well as I fixed an issue with the Extender and updated to Charts 4.4.1 and XDEV 1.1.1
